llvm/test/Analysis/BasicAA/noalias-inttoptr.ll

```diff
@@ -24,21 +24,21 @@ define void @test2(i64 %Q_as_int) {
   ret void
 }
 
-; Verify that escaped noalias parameter may alias inttoptr
+; Verify that escaped noalias parameter are no alias inttoptr
 define void @test3(ptr noalias %P, i64 %Q_as_int) {
   ; CHECK-LABEL: Function: test3:
-  ; CHECK: MayAlias:	i8* %P, i8* %Q
+  ; CHECK: NoAlias:	i8* %P, i8* %Q
```

Comment:
this is not correct. %P and %Q can alias if the caller did ptr2int on %Q.

Reply:
ah, sorry, I missed the noalias bit.
I'm sympathetic to accepting this as valid, but I'm not sure LLVM is ready for that. Probably in most places we only check nolias against other pointer arguments and not against integers.

Reply:
I don't really see how we could accept this, regardless of noalias. Maybe to make it more explicit, we expect that the following code is not UB:

```
test(int * restrict P, uintptr_t Q_as_int) {
    if ((uintptr_t)P == Q_as_int) {
        (int *)Q_as_int = 1; // Can alias P.
    }
}
```

Here `(uintptr_t)P` exposes the provenance of `P` and then `(int *)Q_as_int` chooses it.

(In the test case, the call to `@escape` stands in for the provenance exposure.)

Reply:
That's one interpretation of noalias. But you can also choose to say that P cannot alias with the result of int2ptr since noalias means you cannot do any access through another pointer that points to the same location as P.

Reply:
So you are suggesting that for a `noalias` pointer `%P`, doing `store inttoptr(ptrtoint(%P))` would be UB? I don't think that is viable due to frontend requirements.

Reply:
true, it's not easy to get working. the only solution for now is to lose precision and mark these as may alias.